src/controllers/plugins/api/JSONParse.cpp

```diff
@@ -0,0 +1,379 @@
+#include "controllers/plugins/api/JSONParse.hpp"
+
+#include <boost/json/basic_parser.hpp>
+#include <boost/json/serialize.hpp>
+#include <QVarLengthArray>
+#include <sol/sol.hpp>
+
+namespace {
+
+// NOLINTNEXTLINE(performance-enum-size) -- we want `int` because that's what boost::system::error_code takes
+enum class ParseError : int {
+    MoreThanOneTopLevel = 1,
+    TooMuchNesting,
+    NoElements,
+};
+
+class ParseErrorCategory final : public boost::system::error_category
+{
+public:
+    consteval ParseErrorCategory() = default;
+
+    const char *name() const noexcept override
+    {
+        return "JSON parse error";
+    }
+
+    std::string message(int ev) const override
+    {
+        switch (static_cast<ParseError>(ev))
+        {
+            case ParseError::MoreThanOneTopLevel:
+                return "There was more than one top level element";
+            case ParseError::TooMuchNesting:
+                return "Too much nesting";
+            case ParseError::NoElements:
+                return "Invalid state: no elements on stack";
+        }
+        return "Unknown error code";
+    }
+};
+
+constexpr ParseErrorCategory CATEGORY;
+
+boost::system::error_code makeCode(ParseError kind)
+{
+    return {static_cast<int>(kind), CATEGORY};
+}
+
+// NOLINTBEGIN(cppcoreguidelines-pro-type-vararg)
+// NOLINTBEGIN(readability-identifier-naming)
+// NOLINTBEGIN(readability-convert-member-functions-to-static)
+struct SaxHandler {
+    constexpr static std::size_t max_object_size = 1 << 20;  // about 1 million
+    constexpr static std::size_t max_array_size = 1 << 28;  // about 200 million
+    constexpr static std::size_t max_key_size = 1 << 29;    // 512 MiB
+    constexpr static std::size_t max_string_size = 1 << 30;  // 1 GiB
+
+    constexpr static uint16_t MAX_NESTING = 256;
+
+    SaxHandler(lua_State *state)
+        : state(state)
+    {
+    }
+
+    bool on_document_begin(boost::system::error_code & /* ec */)
+    {
+        return true;
+    }
+
+    bool on_document_end(boost::system::error_code & /* ec */)
+    {
+        return true;
+    }
+
+    bool on_object_begin(boost::system::error_code &ec)
+    {
+        elements.append(Element{
+            .index = 0,
+            .isObject = 1,
+        });
+        if (elements.size() > MAX_NESTING)
+        {
+            ec = makeCode(ParseError::TooMuchNesting);
+            return false;
+        }
+
+        lua_checkstack(this->state, 3);  // grow stack by 3 (table, key, value)
+        lua_newtable(this->state);
+
+        return true;
+    }
+
+    bool on_object_end(size_t /* nElements */, boost::system::error_code &ec)
+    {
+        if (this->elements.empty())
+        {
+            ec = makeCode(ParseError::NoElements);
+            assert(false && "We shouldn't hit this");
```

Comment:
If you want to keep it, make assert text more descriptive. Why do we have an error code for this scenario but we also assert it never happens?